services/galley/src/Galley/API/Update.hs

```diff
@@ -814,6 +811,57 @@ postNewOtrMessage protectee con cnv val msg = do
       gone <- External.deliver toBots
       mapM_ (deleteBot cnv . botMemId) gone
 
+-- | Locally post a message originating from a remote conversation
+-- FUTUREWORK: error handling for missing / mismatched clients
+postRemoteToLocal ::
+  UTCTime ->
+  Remote ConvId ->
+  Qualified UserId ->
+  ClientId ->
+  Maybe Text ->
+  Maybe Priority ->
+  Bool ->
+  UserClientMap Text ->
+  Galley ()
+postRemoteToLocal now conv sender senderc extra priority transient (UserClientMap rcpts) = do
+  localDomain <- viewFederationDomain
+  members <- Data.filterRemoteConvMembers (Map.keys rcpts) (unTagged conv)
```

Comment:
I find it questionable if that is a good idea to do this filtering here. As it stands, we have spread out state across two backends, once the `conversation 1--->* user` mapping on the conversation owning backend, and once the `user 1--->* conversation` mapping on the user owning backend. It's possible these will diverge (and yes we have not yet thought of a good strategy to make them eventually consistent when there are intermittent failures). But in any case there needs to be one source of truth about who the members are of a given conversation, and I'd say that source of truth should lie with the conversation owning backend.

Filtering at this point may not be good idea. Or at least not against this user->conversation table. Possibly this could be filtered against the users table or the clients table (if a user or one of their clients has been deleted, it makes no sense to try and send them a message). But maybe that is assumed to already have happened before message sending starts. In that case maybe no filtering is necessary at all.

Reply:
I think this filtering is necessary given a compromised remote backend may try to send messages to a user who is not supposed to receive messages. I agree that authority on "who is part of a given conversation" is the conversation owning domain. But authority on "whether a user should recieve a message" should always be the user's backend and it should never blindly accept messages for the user.
I also agree that this could lead to weird behaviour when the backends are not on the same page about which users are part of which conversations. But maybe to counter that we can write a log here warning that the conversation has drifted. In future this seems like a good place to initiate some sort of a synchronization.

Reply:
A compromised backend can decide to create a conversation with users from a non-compromised backend inside, and then send them messages (the non-compromised backend will then accept that as nothing gets filtered at that point). So that point to me sounds like a strange scenario.

Logging is a good idea, I agree: we should log at warning level if filtering catches anything, as that shows us that there is either data inconsistency or a compromised backend somewhere.

I still think this precise filter doesn't make sense here; but in the essence of making progress, I'm okay to leave it in at this stage of development, and add a comment on top as a FUTUREWORK to review this as one of the points on the agenda during a general authorization discussion. I propose a `FUTUREWORK(authorization)` for this so we can grep for it.

Reply:
> A compromised backend can decide to create a conversation with users from a non-compromised backend inside, and then send them messages (the non-compromised backend will then accept that as nothing gets filtered at that point). So that point to me sounds like a strange scenario.

This wouldn't be possible when we implement connections.

Reply:
I've added a FUTUREWORK comment as suggested, and a log statement.